source/extensions/transport_sockets/tls/cert_validator/spiffe/spiffe_validator.h

```diff
@@ -30,12 +31,31 @@ namespace Tls {
 
 using X509StorePtr = CSmartPtr<X509_STORE, X509_STORE_free>;
 
-class SPIFFEValidator : public CertValidator {
+struct SpiffeData {
+  absl::flat_hash_map<std::string, CSmartPtr<X509_STORE, X509_STORE_free>> trust_bundle_stores;
+  std::vector<bssl::UniquePtr<X509>> ca_certs;
+
+  int64_t spiffe_refresh_hint;
+  int64_t spiffe_sequence;
+
+  SpiffeData() = default;
+  SpiffeData(const SpiffeData& other) = delete;
+  SpiffeData(SpiffeData&& other) = default;
+};
+
+class SPIFFEValidator : public CertValidator, Logger::Loggable<Logger::Id::secret> {
 public:
-  SPIFFEValidator(SslStats& stats, TimeSource& time_source)
-      : stats_(stats), time_source_(time_source){};
+  SPIFFEValidator(SslStats& stats, Server::Configuration::CommonFactoryContext& context)
+      : tls_(ThreadLocal::TypedSlot<ThreadLocalSpiffeState>::makeUnique(context.threadLocal())),
+        spiffe_data_(std::make_shared<SpiffeData>()),
+        main_thread_dispatcher_(context.mainThreadDispatcher()), stats_(stats),
+        time_source_(context.timeSource()) {
+    tls_->set([](Event::Dispatcher&) { return std::make_shared<ThreadLocalSpiffeState>(); });
+    updateSpiffeData(spiffe_data_);
+  };
```

Comment:
why this change is necessary?

Reply:
tls_ wasn't being initialized in the old constructor, so our tests were segfaulting.

Reply:
I think you can create a simple wrapper method like `OptRef<SpiffeData> getSpiffeData()` in the `SPIFFEValidator` to access the spiffe data.

And if the `tls_` is not created, then means these is no dynamic refresh and we will fallback to use `spiffe_data_ ` directly. Or we will call the tls_->get() to get the data.